server/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -990,21 +991,37 @@ public String getFieldName() {
         }
     }
 
-    private class FieldPath {
+    private static final class FieldPath {
 
-        private final String[] pathElements;
-        private final Object initialContext;
+        private static final int MAX_SIZE = 512;
```

Comment:
I'm wondering if `512` isn't a bit on the high size. That means we'll keep 512 strings (of short to medium length) and 512 `FieldPath`s (each with an array with short strings and a boolean) in memory. Did you happen to check how much heap that occupies in a reasonably "worst" case scenario?

Reply:
The Strings keys are already in memory (they're members of the associated processors), so they're free. The field Paths vary according to the size of the substrings represented (naturally), but for some two segment paths (e.g. `event.original`, `event.message`, etc) I was seeing a retained size of ~140 bytes per FieldPath, so it would cost us 70kb to keep 512 of those around.

I'm fine with dropping that by half or even to a quarter, but beyond that I'd worry that we'd thrash enough to eat into the benefit.

Reply:
Alright, that sounds reasonable to me, thanks :+1:.